tests/test_box_model.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import annotations
+
+
+from textual.box_model import BoxModel, get_box_model
+from textual.css.styles import Styles
+from textual.geometry import Size, Spacing
+
+
+def test_content_box():
+    styles = Styles()
+    styles.width = 10
+    styles.height = 8
+    styles.padding = 1
+    styles.border = ("solid", "red")
+
+    # border-box is default
+    assert styles.box_sizing == "border-box"
+
+    def get_auto_width(container: Size, parent: Size) -> int:
+        assert False, "must not be called"
+
+    def get_auto_height(container: Size, parent: Size) -> int:
+        assert False, "must not be called"
+
+    box_model = get_box_model(
+        styles, Size(60, 20), Size(80, 24), get_auto_width, get_auto_height
+    )
+    # Size should be inclusive of padding / border
+    assert box_model == BoxModel(Size(10, 8), Spacing(0, 0, 0, 0))
+
+    # Switch to content-box
+    styles.box_sizing = "content-box"
+
+    box_model = get_box_model(
+        styles, Size(60, 20), Size(80, 24), get_auto_width, get_auto_height
+    )
+    # width and height have added padding / border to accommodate content
+    assert box_model == BoxModel(Size(14, 12), Spacing(0, 0, 0, 0))
+
+
+def test_width():
+    """Test width settings."""
+    styles = Styles()
+
+    def get_auto_width(container: Size, parent: Size) -> int:
+        return 10
+
+    def get_auto_height(container: Size, parent: Size) -> int:
+        return 10
+
+    box_model = get_box_model(
+        styles, Size(60, 20), Size(80, 24), get_auto_width, get_auto_height
+    )
+    assert box_model == BoxModel(Size(60, 20), Spacing(0, 0, 0, 0))
+
+    # Add a margin and check that it is reported
+    styles.margin = (1, 2, 3, 4)
+
+    box_model = get_box_model(
+        styles, Size(60, 20), Size(80, 24), get_auto_width, get_auto_height
+    )
+    assert box_model == BoxModel(Size(60, 20), Spacing(1, 2, 3, 4))
+
+    # Set width to auto-detect
+    styles.width = "auto"
+
+    box_model = get_box_model(
+        styles, Size(60, 20), Size(80, 24), get_auto_width, get_auto_height
+    )
+    # Setting width to auto should call get_auto_width
+    assert box_model == BoxModel(Size(10, 20), Spacing(1, 2, 3, 4))
+
+    # Set width to 100 vw which should make it the width of the parent
+    styles.width = "100vw"
```

Comment:
This would only be true if the width of the width of the parent is less than the width of the viewport right? If the parent is wider than the viewport then setting the width to 100vw would mean it wouldn't be the width of the parent.

But where does the size of the viewport come into play in these tests? I feel like I'm missing something 🤔

Reply:
The parent here is the viewport, which may be the terminal size, or the dimensions of a scrollable region at the next level up.

Reply:
If the parent wasn't the viewport, wouldn't `get_box_size` need to know what the viewport size is so it can return the correct dimensions?

Reply:
The parent is size is always the viewport. The naming is inherited from where it is called in the layout classes, but I can see how that would be confusing looking at `get_box_model` in isolation. Perhaps at this point the two sizes should be named `container` and `viewport` for clarity.

Reply:
Yeah that would be clearer

Reply:
Done!